# **Recommender Systems**

## **Importing Necessary Libraries and the Dataset**

In [4]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

In [ ]:
movies = pd.read_csv(r"tmdb_5000_movies.csv")
credits = pd.read_csv(r"tmdb_5000_credits.csv")

In [ ]:
movies.head(1)

budget                                             genres   
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  \

                      homepage     id   
0  http://www.avatarmovie.com/  19995  \

                                            keywords original_language   
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en  \

  original_title                                           overview   
0         Avatar  In the 22nd century, a paraplegic Marine is di...  \

   popularity                               production_companies   
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  \

                                production_countries release_date     revenue   
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087  \

   runtime                                   spoken_languages    status   
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released  \

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [ ]:
movies.shape

(4803, 20)

In [ ]:
credits.head(1)

movie_id   title                                               cast   
0     19995  Avatar  [{"cast_id": 242, "character": "Jake Sully", "...  \

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [ ]:
credits.shape

(4803, 4)

The first dataset has the following features:-

* budget - The budget in which the movie was made.
* genre - The genre of the movie, Action, Comedy ,Thriller etc.
* homepage - A link to the homepage of the movie.
* id - This is infact the movie_id as in the first dataset.
* keywords - The keywords or tags related to the movie.
* original_language - The language in which the movie was made.
* original_title - The title of the movie before translation or adaptation.
* overview - A brief description of the movie.
* popularity - A numeric quantity specifying the movie popularity.
* production_companies - The production house of the movie.
* production_countries - The country in which it was produced.
* release_date - The date on which it was released.
* revenue - The worldwide revenue generated by the movie.
* runtime - The running time of the movie in minutes.
* status - "Released" or "Rumored".
* tagline - Movie's tagline.
* title - Title of the movie.
* vote_average -  average ratings the movie recieved.
* vote_count - the count of votes recieved.

The second dataset contains the following features:-

* movie_id - A unique identifier for each movie.
* cast - The name of lead and supporting actors.
* crew - The name of Director, Editor, Composer, Writer etc.

Merging the two dataframes into a single dataframe based on the **Movie Title**

In [ ]:
movies = movies.merge(credits,on='title')

In [ ]:
movies.head(1)

budget                                             genres   
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  \

                      homepage     id   
0  http://www.avatarmovie.com/  19995  \

                                            keywords original_language   
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en  \

  original_title                                           overview   
0         Avatar  In the 22nd century, a paraplegic Marine is di...  \

   popularity                               production_companies  ... runtime   
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0  \

                                    spoken_languages    status   
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released  \

                       tagline   title vote_average vote_count movie_id   
0  Enter the World of Pandora.  Avatar          7.2      11800    19995  \

                                                cast   
0  [{"cast_id": 242, "character": "Jake Sully", "...  \

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 23 columns]

## **Demographic Filtering**

* We need a metric to score or rate movie
* Calculate the score for every movie
* Sort the scores and recommend the best rated movie to the users.

We can use the average ratings of the movie as the score but using this won't be fair enough since a movie with 8.9 average rating and only 3 votes cannot be considered better than the movie with 7.8 as average rating but 40 votes.
So, I'll be using IMDB's weighted rating (wr) which is given as :-

![](https://image.ibb.co/jYWZp9/wr.png)
where,
* v is the number of votes for the movie;
* m is the minimum votes required to be listed in the chart;
* R is the average rating of the movie; And
* C is the mean vote across the whole report

We already have v(**vote_count**) and R (**vote_average**) and C can be calculated as

In [ ]:
C= movies['vote_average'].mean()
C

6.092514036182159

So, the mean rating for all the movies is approx 6 on a scale of 10.
The next step is to determine an appropriate value for m, the minimum votes required to be listed in the chart.
We used the 90th percentile as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 90% of the movies in the list.

In [ ]:
m= movies['vote_count'].quantile(0.9)
m

1842.199999999998

Filtering out the movies that qualify for the chart

In [ ]:
q_movies = movies.copy().loc[movies['vote_count'] >= m]
q_movies.shape

(481, 23)

There are 481 movies which qualify to be in this list. Next, we calculated our metric for each qualified movie.

In [ ]:
def weighted_rating(x, m=m, C=C):
    v = x['vote_count']
    R = x['vote_average']
    # Calculation based on the IMDB formula
    return (v/(v+m) * R) + (m/(m+v) * C)

Defined a new feature 'score' and calculated its value with **`weighted_rating()`**

In [ ]:
q_movies['score'] = q_movies.apply(weighted_rating, axis=1)

In [ ]:
q_movies.head(1)

budget                                             genres   
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  \

                      homepage     id   
0  http://www.avatarmovie.com/  19995  \

                                            keywords original_language   
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en  \

  original_title                                           overview   
0         Avatar  In the 22nd century, a paraplegic Marine is di...  \

   popularity                               production_companies  ...   
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...  \

                                    spoken_languages    status   
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released  \

                       tagline   title vote_average vote_count movie_id   
0  Enter the World of Pandora.  Avatar          7.2      11800    19995  \

                                                cast   
0  [{"cast_id": 242, "character": "Jake Sully", "...  \

                                                crew     score  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  7.050449  

[1 rows x 24 columns]

Finally, we sorted the DataFrame based on the score feature and output the **title, vote count, vote average and score** of the top 10 movies.

In [ ]:
#Sort movies based on score calculated above
q_movies = q_movies.sort_values('score', ascending=False)

In [ ]:
#Print the top 10 movies
q_movies[['title', 'vote_count', 'vote_average', 'score']].head(10)

title  vote_count  vote_average   
1887                       The Shawshank Redemption        8205           8.5  \
662                                      Fight Club        9413           8.3   
65                                  The Dark Knight       12002           8.2   
3237                                   Pulp Fiction        8428           8.3   
96                                        Inception       13752           8.1   
3342                                  The Godfather        5893           8.4   
95                                     Interstellar       10867           8.1   
809                                    Forrest Gump        7927           8.2   
329   The Lord of the Rings: The Return of the King        8064           8.1   
1996                        The Empire Strikes Back        5879           8.2   

         score  
1887  8.058576  
662   7.938689  
65    7.919564  
3237  7.904036  
96    7.862848  
3342  7.850454  
95    7.809015  
809   7.802587  
329   7.726679  
1996  7.697175

Thus, we made the most basic type of recommender that shows the most popular movies

## **Content Based Recommendations**

### Data Pre-processing

**Listing the details(columns) that are not relevant for recommending**
* budget
* homepage
* id
* original_language
* original_title
* popularity
* production_comapny
* production_countries
* release-date(not sure)

Making the dataframe containing only the relevant details

In [ ]:
movies = movies[['movie_id','title','overview','genres','keywords','cast','crew']]

In [ ]:
movies.head()

movie_id                                     title   
0     19995                                    Avatar  \
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview   
0  In the 22nd century, a paraplegic Marine is di...  \
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                              genres   
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  \
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords   
0  [{"id": 1463, "name": "culture clash"}, {"id":...  \
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast   
0  [{"cast_id": 242, "character": "Jake Sully", "...  \
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

Checking are there any null or duplicate values in the dataframe

In [ ]:
movies.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

Since, there are only 3 null entries, we can remove these 3 entries.

In [ ]:
movies.dropna(inplace = True)

In [ ]:
movies.isnull().sum()

movie_id    0
title       0
overview    0
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [ ]:
movies.duplicated().sum()

0

Thus, there are no duplicate entries in the dataset.

### Feature Engineering

In [ ]:
import ast #For converting string into a python list

Function for extracting the genre and keywords from the **'genres'** and **'keywords'** column, which is given in form of a list of dictionary, in which name maps to the required value

In [ ]:
def convert(text):
    L = []
    for i in ast.literal_eval(text):
        L.append(i['name'])
    return L

In [ ]:
movies['genres'] = movies['genres'].apply(convert)
movies.head()

movie_id                                     title   
0     19995                                    Avatar  \
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview   
0  In the 22nd century, a paraplegic Marine is di...  \
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres   
0  [Action, Adventure, Fantasy, Science Fiction]  \
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords   
0  [{"id": 1463, "name": "culture clash"}, {"id":...  \
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast   
0  [{"cast_id": 242, "character": "Jake Sully", "...  \
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [ ]:
movies['keywords'] = movies['keywords'].apply(convert)
movies.head()

movie_id                                     title   
0     19995                                    Avatar  \
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview   
0  In the 22nd century, a paraplegic Marine is di...  \
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres   
0  [Action, Adventure, Fantasy, Science Fiction]  \
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords   
0  [culture clash, future, space war, space colon...  \
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                                cast   
0  [{"cast_id": 242, "character": "Jake Sully", "...  \
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

Extracting the **top 3 cast names** from the **'cast'** column

In [ ]:
movies['cast'] = movies['cast'].apply(lambda x:x[0:3])

In [ ]:
movies.head()

movie_id                                     title   
0     19995                                    Avatar  \
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview   
0  In the 22nd century, a paraplegic Marine is di...  \
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres   
0  [Action, Adventure, Fantasy, Science Fiction]  \
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords cast   
0  [culture clash, future, space war, space colon...  [{"  \
1  [ocean, drug abuse, exotic island, east india ...  [{"   
2  [spy, based on novel, secret agent, sequel, mi...  [{"   
3  [dc comics, crime fighter, terrorist, secret i...  [{"   
4  [based on novel, mars, medallion, space travel...  [{"   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

Function for extracting **directors's name** from the crew list

In [ ]:
def fetch_director(text):
    L = []
    for i in ast.literal_eval(text):
        if i['job'] == 'Director':
            L.append(i['name'])
    return L

In [ ]:
movies['crew'] = movies['crew'].apply(fetch_director)

In [ ]:
movies.head()

movie_id                                     title   
0     19995                                    Avatar  \
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview   
0  In the 22nd century, a paraplegic Marine is di...  \
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres   
0  [Action, Adventure, Fantasy, Science Fiction]  \
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords cast                 crew  
0  [culture clash, future, space war, space colon...  [{"      [James Cameron]  
1  [ocean, drug abuse, exotic island, east india ...  [{"     [Gore Verbinski]  
2  [spy, based on novel, secret agent, sequel, mi...  [{"         [Sam Mendes]  
3  [dc comics, crime fighter, terrorist, secret i...  [{"  [Christopher Nolan]  
4  [based on novel, mars, medallion, space travel...  [{"     [Andrew Stanton]

Function for removing space between words in any name(genre/director/cast)

In [ ]:
def collapse(L):
    L1 = []
    for i in L:
        L1.append(i.replace(" ",""))
    return L1

In [ ]:
movies['cast'] = movies['cast'].apply(collapse)
movies['crew'] = movies['crew'].apply(collapse)
movies['genres'] = movies['genres'].apply(collapse)
movies['keywords'] = movies['keywords'].apply(collapse)

In [ ]:
movies.head()

movie_id                                     title   
0     19995                                    Avatar  \
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview   
0  In the 22nd century, a paraplegic Marine is di...  \
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                         genres   
0  [Action, Adventure, Fantasy, ScienceFiction]  \
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords       cast   
0  [cultureclash, future, spacewar, spacecolony, ...  [[, {, "]  \
1  [ocean, drugabuse, exoticisland, eastindiatrad...  [[, {, "]   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...  [[, {, "]   
3  [dccomics, crimefighter, terrorist, secretiden...  [[, {, "]   
4  [basedonnovel, mars, medallion, spacetravel, p...  [[, {, "]   

                 crew  
0      [JamesCameron]  
1     [GoreVerbinski]  
2         [SamMendes]  
3  [ChristopherNolan]  
4     [AndrewStanton]

Merging the **genres**, **keywords**, **cast** and **crew** column to the **overview** column

In [ ]:
movies['overview'] = movies['overview'].apply(lambda x:x.split())

In [ ]:
movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']

In [ ]:
new_df = movies.drop(columns=['overview','genres','keywords','cast','crew'])

In [ ]:
new_df['tags'] = new_df['tags'].apply(lambda x: " ".join(x))

In [ ]:
new_df.head()

movie_id                                     title   
0     19995                                    Avatar  \
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

### Some more data processing

Converting all the letters to lowercase letters

In [ ]:
new_df['tags'] = new_df['tags'].apply(lambda x:x.lower())

In [ ]:
new_df.head()

movie_id                                     title   
0     19995                                    Avatar  \
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  in the 22nd century, a paraplegic marine is di...  
1  captain barbossa, long believed to be dead, ha...  
2  a cryptic message from bond’s past sends him o...  
3  following the death of district attorney harve...  
4  john carter is a war-weary, former military ca...

Now, we did the stemming of the content in the tags columns

In [ ]:
from nltk.stem.porter import PorterStemmer

In [ ]:
ps = PorterStemmer()

In [ ]:
def stem(text):
    y = []

    for i in text.split():
        y.append(ps.stem(i))
    return " ".join(y)

In [ ]:
new_df['tags'] = new_df['tags'].apply(stem)

In [ ]:
new_df.head()

movie_id                                     title   
0     19995                                    Avatar  \
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  in the 22nd century, a parapleg marin is dispa...  
1  captain barbossa, long believ to be dead, ha c...  
2  a cryptic messag from bond’ past send him on a...  
3  follow the death of district attorney harvey d...  
4  john carter is a war-weary, former militari ca...

### Building the model for recommending the movies

We used the **Bag of Words(BoW)** representation for vectorising the tags

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=5000,stop_words='english')

In [ ]:
vector = cv.fit_transform(new_df['tags']).toarray()

In [ ]:
vector.shape

(4806, 5000)

**Cosine Similarity** was used as a metric for calculating the similarity between the two vectors

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
similarity = cosine_similarity(vector)

In [ ]:
similarity

array([[1.        , 0.08858079, 0.09004503, ..., 0.04559608, 0.        ,
        0.        ],
       [0.08858079, 1.        , 0.06558258, ..., 0.02490677, 0.        ,
        0.0277137 ],
       [0.09004503, 0.06558258, 1.        , ..., 0.02531848, 0.        ,
        0.        ],
       ...,
       [0.04559608, 0.02490677, 0.02531848, ..., 1.        , 0.04003204,
        0.04279605],
       [0.        , 0.        , 0.        , ..., 0.04003204, 1.        ,
        0.08908708],
       [0.        , 0.0277137 , 0.        , ..., 0.04279605, 0.08908708,
        1.        ]])

In [ ]:
similarity.shape

(4806, 4806)

In [ ]:
new_df[new_df['title'] == 'The Lego Movie'].index[0]

744

Function for recommending the movies based on content

In [ ]:
def recommend(movie):
    index = new_df[new_df['title'] == movie].index[0]
    distances = sorted(list(enumerate(similarity[index])),reverse=True,key = lambda x: x[1])
    for i in distances[1:6]:
        print(new_df.iloc[i[0]].title)



Recommending the top 5 similar movies based on the content

In [ ]:
recommend('Gandhi')

Gandhi, My Father
The Wind That Shakes the Barley
Guiana 1838
Mr. Turner
A Passage to India


### Exporting the dataframe and the similarity matrix for web deployement

In [ ]:
import pickle

In [ ]:
pickle.dump(new_df.to_dict(),open('Web_Deployement/movie_list.pkl','wb'))
pickle.dump(similarity,open('Web_Deployement/similarity.pkl','wb'))

## **Collaborative Filtering**

### **Single Value Decomposition(SVD)**

Since the dataset we used before did not have userId(which is necessary for collaborative filtering) we will load another dataset.

We used the **surprise** library to implement SVD.

In [ ]:
#Run this only once to install the package
# !pip install scikit-surprise

In [2]:
from surprise import SVD, Reader
from surprise import Dataset
from surprise.model_selection import cross_validate

In [5]:
reader = Reader()
ratings = pd.read_csv('ratings_small.csv')

In [8]:
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [9]:
ratings.shape

(100004, 4)

In this dataset movies are rated on a scale of 5 unlike the earlier one.

In [10]:
data = Dataset.load_from_df(ratings[['userId', 'movieId', 'rating']], reader)

In [12]:
svd = SVD()
cross_validate(svd, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8997  0.8860  0.8984  0.8994  0.8969  0.8961  0.0051  
MAE (testset)     0.6924  0.6850  0.6919  0.6915  0.6890  0.6900  0.0027  
Fit time          1.22    0.93    0.92    1.22    0.97    1.05    0.14    
Test time         0.24    0.10    0.10    0.21    0.09    0.15    0.06    


{'test_rmse': array([0.89974482, 0.88604812, 0.89843677, 0.89939118, 0.8969227 ]),
 'test_mae': array([0.69241038, 0.68501343, 0.6919253 , 0.69145977, 0.68902418]),
 'fit_time': (1.223158359527588,
  0.933720588684082,
  0.9231717586517334,
  1.217860221862793,
  0.9667320251464844),
 'test_time': (0.2376558780670166,
  0.09539175033569336,
  0.0992588996887207,
  0.21053123474121094,
  0.0868985652923584)}

We get a **Root Mean Sqaure Error(RMSE) of 0.89** approx which is more than good enough for our case.
We will now train on our dataset and arrive at predictions.

In [13]:
trainset = data.build_full_trainset()
svd.fit(trainset)

For instance, we pick user with user Id - 1  and check the ratings he/she has given.

In [14]:
ratings[ratings['userId'] == 1]

userId  movieId  rating   timestamp
0        1       31     2.5  1260759144
1        1     1029     3.0  1260759179
2        1     1061     3.0  1260759182
3        1     1129     2.0  1260759185
4        1     1172     4.0  1260759205
5        1     1263     2.0  1260759151
6        1     1287     2.0  1260759187
7        1     1293     2.0  1260759148
8        1     1339     3.5  1260759125
9        1     1343     2.0  1260759131
10       1     1371     2.5  1260759135
11       1     1405     1.0  1260759203
12       1     1953     4.0  1260759191
13       1     2105     4.0  1260759139
14       1     2150     3.0  1260759194
15       1     2193     2.0  1260759198
16       1     2294     2.0  1260759108
17       1     2455     2.5  1260759113
18       1     2968     1.0  1260759200
19       1     3671     3.0  1260759117

In [15]:
prediction = svd.predict(1, 302, 3)

In [16]:
print(prediction.est)

2.655764658451391


For movie with ID 302, we get an estimated prediction of **2.618**.

One startling feature of this recommender system is that it doesn't care what the movie is (or what it contains). It works purely on the basis of an assigned movie ID and tries to predict ratings based on how the other users have predicted the movie.